base/compiler/ssair/passes.jl

```diff
@@ -1082,22 +1087,74 @@ within `sroa_pass!` which redirects references of `typeassert`ed value to the co
 function adce_pass!(ir::IRCode)
     phi_uses = fill(0, length(ir.stmts) + length(ir.new_nodes))
     all_phis = Int[]
+    unionphis = Int[] # sorted
+    unionphi_types = Any[]
     compact = IncrementalCompact(ir)
     for ((_, idx), stmt) in compact
         if isa(stmt, PhiNode)
             push!(all_phis, idx)
-        elseif is_known_call(stmt, typeassert, compact) && length(stmt.args) == 3
-            # nullify safe `typeassert` calls
-            ty, isexact = instanceof_tfunc(argextype(stmt.args[3], compact))
-            if isexact && argextype(stmt.args[2], compact) ⊑ ty
-                compact[idx] = nothing
+            if isa(compact.result[idx][:type], Union)
+                push!(unionphis, idx)
+                push!(unionphi_types, Union{})
+            end
+        elseif isa(stmt, PiNode)
+            val = stmt.val
+            if isa(val, SSAValue) && is_union_phi(compact, val.id)
+                r = searchsorted(unionphis, val.id)
+                if !isempty(r)
+                    unionphi_types[first(r)] = Union{unionphi_types[first(r)], widenconst(stmt.typ)}
+                end
+            end
+        else
+            if is_known_call(stmt, typeassert, compact) && length(stmt.args) == 3
+                # nullify safe `typeassert` calls
+                ty, isexact = instanceof_tfunc(argextype(stmt.args[3], compact))
+                if isexact && argextype(stmt.args[2], compact) ⊑ ty
+                    compact[idx] = nothing
+                    continue
+                end
+            end
+            for ur in userefs(stmt)
+                use = ur[]
+                if isa(use, SSAValue) && is_union_phi(compact, use.id)
+                    r = searchsorted(unionphis, use.id)
+                    if !isempty(r)
+                        deleteat!(unionphis, first(r))
+                        deleteat!(unionphi_types, first(r))
+                    end
+                end
             end
         end
     end
     non_dce_finish!(compact)
     for phi in all_phis
         count_uses(compact.result[phi][:inst]::PhiNode, phi_uses)
     end
+    # Narrow any union phi nodes that have unused branches
+    for (phi, t) in zip(unionphis, unionphi_types)
```

Comment:
I remember @vtjnash said that this `(phi, t)` may allocate arbitrary tuple types depending on user types fed into `unionphi_types`:
```suggestion
    @assert length(unionphis) == length(unionphi_types)
    for i = length(unionphis)
        phi = unionphis[i]
        t = unionphi_types[i]
```

Reply:
Ah, right. Good catch.